server/src/main/java/org/elasticsearch/index/engine/InternalEngine.java

```diff
@@ -2731,7 +2734,11 @@ private boolean assertMaxSeqNoOfUpdatesIsAdvanced(Term id, long seqNo, boolean a
     private void restoreVersionMapAndCheckpointTracker(DirectoryReader directoryReader) throws IOException {
         final IndexSearcher searcher = new IndexSearcher(directoryReader);
         searcher.setQueryCache(null);
-        final Query query = LongPoint.newRangeQuery(SeqNoFieldMapper.NAME, getPersistedLocalCheckpoint() + 1, Long.MAX_VALUE);
+        final Query query = new BooleanQuery.Builder()
+            .add(LongPoint.newRangeQuery(
+                    SeqNoFieldMapper.NAME, getPersistedLocalCheckpoint() + 1, Long.MAX_VALUE), BooleanClause.Occur.MUST)
+            .add(Queries.newNonNestedFilter(), BooleanClause.Occur.MUST) // exclude non-root nested documents
+            .build();
         final Weight weight = searcher.createWeight(query, ScoreMode.COMPLETE_NO_SCORES, 1.0f);
```

Comment:
Lucene requires queries to be rewritten before creating a weight. This particular boolean query is fine because it rewrites to itself, but it'd be more future-proof to call rewrite().

```suggestion
         final Query rewrittenQuery = searcher.rewrite(query);
         final Weight weight = searcher.createWeight(rewrittenQuery, ScoreMode.COMPLETE_NO_SCORES, 1.0f);
```

Reply:
Yes, I pushed https://github.com/elastic/elasticsearch/pull/51279/commits/8c37fb02817e06c78add9c3b7d8ca8f0e472f82a.